src/email/email_provider.rs

```diff
@@ -1,41 +1,70 @@
+use super::mail_cmd::MailCommand;
 use super::sendgrid::SendGrid;
 use enum_dispatch::enum_dispatch;
-use log::warn;
 
 /// An email provider abstraction to allow for multiple backends.
 #[enum_dispatch]
 pub trait EmailProvider {
   /// Sends an email to and from the specified address.
-  fn send_email(&self, address: &str, api_key: &str, contents: &str);
+  fn send_email(&self, address: &str, contents: &str);
 }
 
 /// An enum containing all Email Provider implementations.
 #[enum_dispatch(EmailProvider)]
 enum EmailProviders {
   SendGrid(SendGrid),
+  MailCommand(MailCommand),
 }
 
-impl From<String> for EmailProviders {
-  fn from(input: String) -> Self {
-    // Note that the input is trimmed and converted
-    // to upper case for the sake of consistency
-    match input.trim().to_uppercase().as_str() {
-      "SENDGRID" => Self::SendGrid(SendGrid::new()),
-      e => {
-        warn!("Invalid Email provider: {}, defaulting to SendGrid.", e);
-        Self::SendGrid(SendGrid::new())
+impl TryFrom<String> for EmailProviders {
+  type Error = String;
+
+  fn try_from(value: String) -> Result<Self, Self::Error> {
+    let client = value.trim().to_uppercase();
+    match client.as_str() {
+      "SENDGRID" => {
+        let api_key = std::env::var("API_KEY");
+        if let Ok(the_key) = api_key {
+          Ok(Self::SendGrid(SendGrid::new(the_key)))
+        } else {
+          Err("Cannot use SendGrid without API_KEY".to_owned())
+        }
```

Comment:
This is more of an open-ended question than a "pls fix" but would it make more sense to move this logic inside the SendGrid implementation somehow? I feel like it would make more sense to have the implementations care about their specific edge cases than the provider.

Reply:
I tend to agree with the point.
What I can say is that: having internal modules less intertwined with ENV variables and stuff make them easier to test, re-implement and extend.
Makes them more unitary.

The application layer (main.rs and email_provider.rs) are dealing with how user parameters reach your core logic, this is fine as a design principle.

You might want to load the API_KEY from somewhere else, one day, and this can be achieved by writing a different `email/email_provider.rs` that just uses the `email/` modules in a different way.

Reply:
I'll look into this before merging. If I come up with any ideas I will likely post them in the PR thread before implementing them.